[im/subscribe] Enable subscription over interaction model protocol

erjiaqing opened the pull request:
#### Problem
The current "report" routine is not spec compliance.

#### Change overview

- Update templates to use interaction model APIs for sending subscribe
requests.
- Update DeviceControllerInteractionModelDelegate to handle report
messages from subscription.
- Update ChipCallback Mgr to bridge subscribe responses to existing
callbacks, and handle TLV message just as read responses.


#### Testing

- Update TestSuite for basic Subscription protocol tests, test routing
looks like follows:
  - Send subscribe request, wait for response.
  - Wait for a few seconds, and execute "kick" commands.
  - The test is expected to receive the same number of Reports as the
  number of kick commands.
  - The test will fail when not receiving enough report data from server
  in global timeout.
- Update python script test to (in cirque) include a similar test
routine as test suite.

erjiaqing committed:
[im/subscribe] Enable subscription over interaction model protocol

- Update templates to use interaction model APIs for sending subscribe
requests.
- Update DeviceControllerInteractionModelDelegate to handle report
messages from subscription.
- Update ChipCallback Mgr to bridge subscribe responses to existing
callbacks, and handle TLV message just as read responses.
- Update TestSuite for basic Subscription protocol tests, test routing
looks like follows:
  - Send subscribe request, wait for response.
  - Wait for a few seconds, and execute "kick" commands.
  - The test is expected to receive the same number of Reports as the
  number of kick commands.
  - The test will fail when not receiving enough report data from server
  in global timeout.
- Update python script test to (in cirque) include a similar test
routine as test suite.

asakhart commented on src/app/reporting/reporting.cpp:
```diff
@@ -694,53 +694,6 @@ EmberStatus emAfPluginReportingRemoveEntry(uint8_t index)
     return status;
 }
 
-void emberAfReportingAttributeChangeCallback(EndpointId endpoint, ClusterId clusterId, AttributeId attributeId, uint8_t mask,
```
Please do not include old code cleanup until later changes after we get this landed and only focus on the core of this change to make the delta to enable subs easier to process.

When we go to share, we should also try to upload this change on top of a merged PR from Yunhan on a branch and do code change as one commit, followed by test changes as another commit to ease looking at code delta.

erjiaqing replied:
This is the only (and necessary) removal to get code compiled, or we will have two `emberAfReportingAttributeChangeCallback` and linker will be unhappy.

asakhart replied:
Yes, but why do you have to replicate the ember name? I gather something is calling that function an that's why you are just replacing it so the linker picks up your version, but couldn't you name your function differently and replace the name at the call site of emberAfReportingAttributeChangeCallback? Is that call site not available to you?

erjiaqing committed:
Run Codegen

yunhanw-google committed:
add multiple tests

--Testing
Add unit test for positive and negative subscribe test
Add integrated cirque test for subscribe

erjiaqing committed:
Change bridge-app call

andy31415 commented on src/darwin/Framework/CHIP/templates/clusters-tests.zapt:
```diff
@@ -172,40 +172,7 @@ CHIPDevice * GetPairedDevice(uint64_t deviceId)
-    }];
-    [self waitForExpectationsWithTimeout:kTimeoutInSeconds handler:nil];
-}
+// Spec compliant reporting Tests goes into test suite now.
```
comment probably not needed - I would just leave the delete.
